### ISSUE TITLE
twisted.spread.util.FilePager keeps all sent chunks in memory

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1843]
|<img alt="ino's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ino reported|
|-|-|
|Trac ID|trac#1843|
|Type|defect|
|Created|2006-06-22 13:57:16Z|

When sending a file, t.s.u.FilePager stores each chunk read (from the producer t.p.b.FileSender) in a list. This becomes a problem when the file being sent is very large, as it uses up a lot of memory.

A simple patch for this is attached. This patch pops from the chunks-list when sending, so that sent data isn't kept in memory.

Attachments:

* [pop-chunks-1843.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c77/c77be4d776f1eb0b0653d3d825d5cedb6950c1e9/372c6cf82b109c05fad10c9b95dfa59201ccf8ab.diff) (813 bytes) - added by ino on 2006-06-22 14:00:07Z - pop chuncks when sending

<details><summary>Searchable metadata</summary>

```
trac-id__1843 1843
type__defect defect
reporter__ino ino
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1150984636000000 1150984636000000
changetime__1182673040000000 1182673040000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

This bugfix needs an accompanying unit test.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

It's ready to review in pb-pager-1843+2321. I did a lot of cleanups because test_pb was a mess, so I hope it won't perturb review too much. I can split this if necessary, but it has to be done I think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

  * twisted.spread.util.FilePager.sendNextPage should have a docstring
  * twisted.test.test_pb.createFactoryCopy should have a docstring
  * NewStyleTestCase.setUp and .tearDown should have docstrings
  * If NewStyleTestCase.setUp called getRootObject, added a callback to save the root as an attribute of itself, and returned the Deferred, each of the test methods would be simplified somewhat.
  * NewStyleTestCase.test_newStyle and .test_alloc should have docstrings.  Probably all the callback methods should as well, but you didn't modify them so you don't have to add them.
  * NewStyleCachedTestCase.test_newStyleCache should have a docstring
  * Now I know about ```twisted.spread.publish``` and I didn't before :(
  * The stuff moved out of test_spread should have docstrings.  You didn't really change it, so you probably don't have to add them in this branch, but svn blame will say you wrote this undocumented code now, if you don't add them. ;)

The sendNextPage change and test look good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

I think I've addressed everything, and tried to document the most I could. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

Looks great, please merge

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [20609]) Merge pb-pager-1843+2321

Authors: ino, therve
Reviewer: exarkun
Fixes [#1843](https://github.com/twisted/twisted/issues/1843)
Fixes [#2321](https://github.com/twisted/twisted/issues/2321)

Modify FilePager so that data read is not kept in memory. In the process,
do a lot of cleanups in test_pb to match current coding standards.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
